### PR TITLE
server: maintain chat completion id for streaming responses

### LINKED ISSUE #5876: Consistent chat completion id in OpenAI compatible chat completion endpoint
First thank you for this great project!

I was wondering if for [OpenAI compatible chat completion endpoint](https://github.com/ggerganov/llama.cpp/blob/master/examples/server/README.md#result-json-1), the streaming responses should return the same completion id (`chatcmpl-`) for each chunk.

For example the chat completion ids are different (`chatcmpl-2EHCQqsRzdOlFskNehCMu2oOMTXhSjey`, `chatcmpl-Cm7q0Ru5uEGVlW4r6cZaGNQrlS7oF724`) in the following response:
```
{"choices":[{"delta":{"content":"Under"},"finish_reason":null,"index":0}],"created":1709591108,"id":"chatcmpl-2EHCQqsRzdOlFskNehCMu2oOMTXhSjey","model":"gpt-3.5-turbo","object":"chat.completion.chunk"}

{"choices":[{"delta":{"content":"stood"},"finish_reason":null,"index":0}],"created":1709591108,"id":"chatcmpl-Cm7q0Ru5uEGVlW4r6cZaGNQrlS7oF724","model":"gpt-3.5-turbo","object":"chat.completion.chunk"}

...
```

Example NodeJS code that generates the above chunks:
```ts
import OpenAI from "openai";

process.env["OPENAI_API_KEY"] =
  "no-key";

const openai = new OpenAI({
  baseURL: "http://127.0.0.1:8080/v1",
  apiKey: "no-key",
});

async function main() {
  const stream = await openai.chat.completions.create({
    model: "gpt-3.5-turbo",
    messages: [{ role: "user", content: "Say this is a test" }],
    stream: true,
  });
  for await (const chunk of stream) {
    process.stdout.write(JSON.stringify(chunk));
  }
}

main();
```

This is fine for nodejs server side generation, but if I stream the HTTP response and consume it with OpenAI's NodeJS SDK, I would get this error: `missing finish_reason for choice 0`. It seems that when a different chunk id is supplied [here](https://github.com/openai/openai-node/blob/6175eca426b15990be5e5cdb0e8497e547f87d8a/src/lib/ChatCompletionStream.ts#L129-L137), `#endRequest` is called prematurely and the correspondent chunk would not have a `finish_reason`.

[Example client side code](https://github.com/openai/openai-node/blob/6175eca426b15990be5e5cdb0e8497e547f87d8a/examples/stream-to-client-browser.ts#L11):
```ts
import fetch from 'node-fetch';
import { ChatCompletionStream } from 'openai/lib/ChatCompletionStream';

fetch('http://localhost:3000', {
  method: 'POST',
  body: 'Tell me why dogs are better than cats',
  headers: { 'Content-Type': 'text/plain' },
}).then(async (res) => {
  // @ts-ignore ReadableStream on different environments can be strange
  const runner = ChatCompletionStream.fromReadableStream(res.body);

  runner.on('content', (delta, snapshot) => {
    process.stdout.write(delta);
    // or, in a browser, you might display like this:
    // document.body.innerText += delta; // or:
    // document.body.innerText = snapshot;
  });

  console.dir(await runner.finalChatCompletion(), { depth: null });
});
```

### PR DESCRIPTION
**Pull request by mscheong01**
fixes #5876 

tested code ( provided by @xyc )
```javascript
import OpenAI from "openai";

process.env["OPENAI_API_KEY"] =
  "no-key";

const openai = new OpenAI({
  baseURL: "http://127.0.0.1:8080/v1",
  apiKey: "no-key",
});

async function main() {
  const stream = await openai.chat.completions.create({
    model: "gpt-3.5-turbo",
    messages: [{ role: "user", content: "Say this is a test" }],
    stream: true,
  });
  for await (const chunk of stream) {
    process.stdout.write(JSON.stringify(chunk));
  }
}

main();
```

result (before)
```
{"choices":[{"delta":{"content":"S"},"finish_reason":null,"index":0}],"created":1709628034,"id":"chatcmpl-LMLcrWfJ29GgNvzP7tYhSSC32xARD6p1","model":"gpt-3.5-turbo","object":"chat.completion.chunk"}{"choices":[{"delta":{"content":"ay"},"finish_reason":null,"index":0}],"created":1709628034,"id":"chatcmpl-B5x5NUWQhvUCrh4oPQkigOfFvgdEe6kM","model":"gpt-3.5-turbo","object":"chat.completion.chunk"}{"choices":[{"delta":{"content":" this"},"finish_reason":null,"index":0}],"created":1709628046,"id":"chatcmpl-34ohfUOtEixbtgGVJY9CIkPFZ1BCwC3o","model":"gpt-3.5-turbo","object":"chat.completion.chunk"}{"choices":[{"delta":{"content":" is"},"finish_reason":null,"index":0}],"created":1709628048,"id":"chatcmpl-mk5o2qKT46tvoShPZ2LNr3RjiDupMIGt","model":"gpt-3.5-turbo","object":"chat.completion.chunk"}{"choices":[{"delta":{"content":" a"},"finish_reason":null,"index":0}],"created":1709628049,"id":"chatcmpl-6RXeg13CV7h1jhDDsbTUfoy0nysh67iq","model":"gpt-3.5-turbo","object":"chat.completion.chunk"}{"choices":[{"delta":{"content":" test"},"finish_reason":null,"index":0}],"created":1709628050,"id":"chatcmpl-hnlmLdKAvr0YV0eXUcIfc6uLYDdHkKXV","model":"gpt-3.5-turbo","object":"chat.completion.chunk"}{"choices":[{"delta":{},"finish_reason":"stop","index":0}],"created":1709628073,"id":"chatcmpl-zTcUBKVba00yAd0PylbJJFEcbqJr5RXs","model":"gpt-3.5-turbo","object":"chat.completion.chunk"}
```

result (after)
```
{"choices":[{"delta":{"content":"S"},"finish_reason":null,"index":0}],"created":1709627698,"id":"chatcmpl-OvKzfgQykk3jjHM2liQCrwnCwzJWKUAk","model":"gpt-3.5-turbo","object":"chat.completion.chunk"}{"choices":[{"delta":{"content":"ay"},"finish_reason":null,"index":0}],"created":1709627706,"id":"chatcmpl-OvKzfgQykk3jjHM2liQCrwnCwzJWKUAk","model":"gpt-3.5-turbo","object":"chat.completion.chunk"}{"choices":[{"delta":{"content":" this"},"finish_reason":null,"index":0}],"created":1709627707,"id":"chatcmpl-OvKzfgQykk3jjHM2liQCrwnCwzJWKUAk","model":"gpt-3.5-turbo","object":"chat.completion.chunk"}{"choices":[{"delta":{"content":" is"},"finish_reason":null,"index":0}],"created":1709627708,"id":"chatcmpl-OvKzfgQykk3jjHM2liQCrwnCwzJWKUAk","model":"gpt-3.5-turbo","object":"chat.completion.chunk"}{"choices":[{"delta":{"content":" a"},"finish_reason":null,"index":0}],"created":1709627709,"id":"chatcmpl-OvKzfgQykk3jjHM2liQCrwnCwzJWKUAk","model":"gpt-3.5-turbo","object":"chat.completion.chunk"}{"choices":[{"delta":{"content":" test"},"finish_reason":null,"index":0}],"created":1709627710,"id":"chatcmpl-OvKzfgQykk3jjHM2liQCrwnCwzJWKUAk","model":"gpt-3.5-turbo","object":"chat.completion.chunk"}{"choices":[{"delta":{},"finish_reason":"stop","index":0}],"created":1709627727,"id":"chatcmpl-OvKzfgQykk3jjHM2liQCrwnCwzJWKUAk","model":"gpt-3.5-turbo","object":"chat.completion.chunk"}
```

### CODE REVIEW

File: examples/server/server.cpp
Comment (ggerganov) [inline examples/server/server.cpp:3213]: This seems to make all responses have the same id.

Don't we want to have a separate id for separate streams?

Comment (mscheong01) [reply examples/server/server.cpp]: Yes, I messed up 😞 Can confirm that [aa95dc5](https://github.com/ggerganov/llama.cpp/pull/5880/commits/aa95dc5568f9a27f17816426c3a8c118fc7789e3) fixes this up

**Review by ngxson — COMMENTED** [review]
This PR is LGTM, one thing that would be nice if you can do:

We're having multiple places calling `gen_chatcmplid`. It would be nice to call `gen_chatcmplid` only once for each incoming request, then use the generated id in both `format_final_response_oaicompat` and `format_partial_response_oaicompat`

Just small detail though, we can do it later. I'm planning to clean up all the functions related to OAI-compat.

**Comment by mscheong01** [comment]
> We're having multiple places calling gen_chatcmplid. It would be nice to call gen_chatcmplid only once for each incoming request, then use the generated id in both format_final_response_oaicompat and format_partial_response_oaicompat

That was how my initial draft was written like, but for some reason(probably because the random string generator is not thread safe) the generated completion_ids were garbage value like:
i.e)
```
"id":"Pְ�ug\u0000\u0000�\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000"
```
So I had to come up with the current version.
I'll try fixing this if possible before this PR gets merged.

**Review by ngxson — APPROVED** [review]
This LGTM, thank you. I'll wait for @ggerganov to decide if this can be merged now or after #5882

**Comment by ggerganov** [comment]
Thanks for taking a look. Will be reimplemented in #5882 - no need to merge this PR

**Comment by xyc** [comment]
Thank you for the quick turnaround, @mscheong01 @ggerganov .

I don't have much to add - it works fine and generates consistent chat completion id. Only issue is when consumed by OpenAI client (as in the code second block of https://github.com/ggerganov/llama.cpp/issues/5876#issue-2167941634) it seems that I would still get an error (`missing role for choice 0`)

Adding `{"role", "assistant"},` before these lines seems to fix it.

https://github.com/ggerganov/llama.cpp/blob/293378b20bce8c65b52ac7b784f2aa1d55ec3256/examples/server/oai.hpp#L176

https://github.com/ggerganov/llama.cpp/blob/293378b20bce8c65b52ac7b784f2aa1d55ec3256/examples/server/oai.hpp#L197

**Comment by ggerganov** [comment]
@mscheong01 If you can re-apply the changes on top of latest `master` we can merge this. Will close this for now

**Pull request closed without merging**
